Issue #5023: Fully Parallel Partitioning

hawkfish opened the pull request:
Create a queue for all the sorting tasks.
This lets us sort more than one partition at a time.

Commit:
Issue #5023: Fully Parallel Sorting

Create a queue for all the sorting tasks.
This lets us sort more than one partition at a time.

lnkuiper commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -1483,34 +1483,203 @@ unique_ptr<GlobalSinkState> PhysicalWindow::GetGlobalSinkState(ClientContext &co
+
+	switch (stage) {
+	case SortStage::PREPARE:
+		sort_state.PrepareMergePhase();
```
Are all threads busy during this `Event`?

I am a bit concerned about calling `PrepareMergePhase` while holding the lock. Other threads may have to wait until the thread holding the lock has completed `PrepareMergePhase`, even though they could be preparing the merge phase for other sorts.

Would it be possible to assign a `PREPARE` task to a thread, then drop the lock, and then the thread can do the preparation lock-free? If all threads are busy and this is not a problem, feel free to ignore this.

As a side note, I am looking at the code for `PrepareMergePhase`. This re-evaluates whether we need to be doing an external sort, by checking whether the variable-size data fits in memory. If it does, it unswizzles pointers so that the variable-size data does not have to be moved during the merge sort. However, if `external` is already `true` at this time, no unswizzling happens, and `PrepareMergePhase` is not costly at all. This check is not accurate for the window operator, because you have more than one sort going on at a time.

Not something you have to fix in this PR, but something to keep in mind.

hawkfish replied:
Yes I'm concerned about the relatively high levels of lock contention I was observing, especially near the end (the 'find a hash group" loop is N^2 in the numbers of groups.). 

Which operations are slow? I somehow got the impression that they were all fast except the merging, but now that I think about it, I got started down this road by noticing that unswizzling during prepare was time-consuming.

lnkuiper replied:
I think the pattern here is to only grab the lock to assign a task / mark a task as done, and perform all other work lock-free. However, we should look into why PrepareMergePhase is always unswizzling. This should only happen if the data is close to the memory limit.

hawkfish replied:
Yeah the new PR does this.  And my test is rather brutal (Fannie Mae SF 4 uses about 70GB)

lnkuiper reviewed (approved):
Great that you got this to work so quickly! Fully sorting all hash groups in a single event is much cleaner than the chained events that we have in `PhysicalOrder` (something I might want to change in the future).

I have one remark about `PrepareMergePhase`, but that is mostly writing some thoughts down that should be considered in a future PR.

Commit:
Merge branch 'master' into hawkfish-window-masks

Commit:
Issue #5023: Window Prepare Parallelisation

Move the PrepareMergePhase calls into tasks
to avoid holding TryPrepareNextState locks..

lnkuiper commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -1623,17 +1640,20 @@ class WindowMergeTask : public ExecutorTask {
 
 		// Thread is done with its assigned task, try to fetch new work
-		for (auto &global_state : hash_groups.states) {
+		for (auto group = sorted; group < hash_groups.states.size(); ++group) {
```
Does it always work to start the loop at `group = sorted`?

What if the first hash group is large, let's say it can occupy 3 threads, and the second hash group is small; it can occupy 1 thread. Assuming we have 4 threads, a thread will start working on the second hash group before the first hash group is completed.

The second hash group is small enough that it will be completed before the first hash group. Now, `sorted` is updated from `0` to `1`, and we will never complete sorting the first hash group.

Does this make sense? Or perhaps I am overlooking something.

lnkuiper reviewed (commented):
Just looked at your changes, and I have some comments:

Commit:
Issue #5023: Window Prepare Parallelisation

Fix stage race condition.
Clarify low water mark calculation.

Commit:
Merge branch 'master' into hawkfish-window-masks

hawkfish commented on src/execution/operator/aggregate/physical_window.cpp:
```diff
@@ -1502,34 +1504,43 @@ class WindowLocalMergeState {
-	atomic<idx_t> total_tasks;
-	atomic<idx_t> tasks_assigned;
-	atomic<idx_t> tasks_completed;
+	WindowSortStage stage;
```
Ah I misunderstood where the race was. This is easy to fix by putting the lock in `IsSorted`.

Commit:
Issue #5023: Window Prepare Parallelisation

Really fix stage race condition...

Commit:
Merge branch 'master' into hawkfish-window-masks

hawkfish commented:
The race condition is gone and I think @Mytherin just fixed the ODBC build issue. Of the two benchmark failures, the micro one does not involve windowing and is likely spurious; the TPC-DS one does not reproduce on my machine (in fact it clearly got about 10% faster and more stable). The code coverage test crash does not involve windowing and I can't reproduce it.

Mytherin commented:
Thanks! LGTM
